### PR TITLE
dateutils: improvements to date getting [WIP]

### PR DESCRIPTION
**Pull request by bittirousku**
The improved date getter/formatter.  I have tested it here with Elsevier and BASE spiders.

* How should the code be organised? Right now everything's happening in `dateutils.py`.

* Should the date formatting happen in the spider or in the loader? At least with `elsevier_spider`
  it's more convenient to be inside the spider as I'm getting the year at the same time.

### CODE REVIEW

**Comment by jalavik** [comment]
I think in general the date checking and parsing can happen in the loader as it is then applied equally everywhere.

If you need the year in a spider I would just parse it individually in the spider I think.

**Comment by bittirousku** [comment]
Ok, as the year is extracted from the formatted date, the formatting function would be called twice. Once in the spider, once in the loader. Is that ok?

**Comment by jalavik** [comment]
sure

File: hepcrawl/spiders/elsevier_spider.py [hunk 1]
```diff
@@ -797,18 +812,20 @@ def parse_node(self, response, node):
...
-            return request
+            try:
+                request = Request(sd_url, callback=self.scrape_sciencedirect)
+                request.meta["info"] = info
+                request.meta["keys_missing"] = keys_missing
+                request.meta["node"] = node
+                request.meta["xml_url"] = xml_file
+                request.meta["handle_httpstatus_list"] = self.ERROR_CODES
+                return request
+            except ValueError:
```
Comment (jalavik) [inline hepcrawl/spiders/elsevier_spider.py]: Can you add a comment here explaining the reason for catching this potential exception?

File: hepcrawl/spiders/elsevier_spider.py [hunk 2]
```diff
@@ -548,24 +544,37 @@ def _get_ref_volume(ref):
         return ", ".join(volumes)
 
     @staticmethod
-    def _get_ref_years(ref):
-        """Get the reference year.
+    def range_as_string(data):
```
Comment (jalavik) [inline hepcrawl/spiders/elsevier_spider.py]: Should perhaps this function be moved to utils and get a couple of tests?

**Comment by jalavik** [comment]
Looks good. Thanks for this! Will merge when the minor comments are addressed. Just ping me.

**Comment by bittirousku** [comment]
@jalavik I made some changes!

**Comment by jalavik** [comment]
:+1:

**Pull request merged**